### ISSUE TITLE
Rule Change: (wrap-iife)

### ISSUE THREAD

**Issue by fab1o** [eslint/eslint#15874]
### What rule do you want to change?

wrap-iife

### What change to do you want to make?

Generate more warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```js
((a) => {
    return a === 'a';
})('');
```

### What does the rule currently do for this code?

Nothing

### What will the rule do after it's changed?

Not allow

### Participation

- [ ] I am willing to submit a pull request to implement this change.

### Additional comments

It seems that this rule only applies for non-arrow functions. My suggestion is that it should apply for arrow function as well. Or at least have an options for it.

**Comment by fab1o** [comment]
What's the rule for not allow self-invoking functions of any kind? That's what I'm looking for... Thanks!

**Comment by mdjermanovic** [comment]
Hi @fab1o, thanks for the issue!

The wrap-iife rule doesn't disallow IIFEs, it just requires additional parentheses.

```js
var x = function(){}(); // disallowed

var x = (function(){}()); // allowed
var x = (function(){})(); // allowed depending on the configuration
```

The rule doesn't apply to arrow functions because the syntax already requires them to be parenthesized when they are immediately invoked.

```js
var x = () => {} (); // SyntaxError: Unexpected token '('
```

> What's the rule configuration for not allow self-invoking functions of any kind? That's what I'm looking for... Thanks!

We don't have a core rule that disallows IIFEs, but you can configure the [no-restricted-syntax](https://eslint.org/docs/rules/no-restricted-syntax) rule for this purpose:

```js
/* eslint no-restricted-syntax: [
    "error",
    {
        "selector": "CallExpression > :matches(FunctionExpression, ArrowFunctionExpression).callee",
        "message": "IIFEs are not allowed."
    }
]*/

((a) => {
    return a === 'a';
})(''); // error

(function(){}()) // error
```

[Online Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXJlc3RyaWN0ZWQtc3ludGF4OiBbXG4gICAgXCJlcnJvclwiLFxuICAgIHtcbiAgICAgICAgXCJzZWxlY3RvclwiOiBcIkNhbGxFeHByZXNzaW9uID4gOm1hdGNoZXMoRnVuY3Rpb25FeHByZXNzaW9uLCBBcnJvd0Z1bmN0aW9uRXhwcmVzc2lvbikuY2FsbGVlXCIsXG4gICAgICAgIFwibWVzc2FnZVwiOiBcIklJRkVzIGFyZSBub3QgYWxsb3dlZC5cIlxuICAgIH1cbl0qL1xuXG4oKGEpID0+IHtcbiAgICByZXR1cm4gYSA9PT0gJ2EnO1xufSkoJycpOyAvLyBlcnJvclxuXG4oZnVuY3Rpb24oKXt9KCkpIC8vIGVycm9yXG4iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOiJsYXRlc3QiLCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7fSwiZW52Ijp7fX19)
